server/matchmaker/algorithm/team_matchmaker.py

```diff
@@ -0,0 +1,262 @@
+import statistics as stats
+from collections import defaultdict
+from typing import Dict, Iterable, List, Tuple
+
+from ...config import config
+from ...decorators import with_logger
+from ..search import CombinedSearch, Game, Match, Search
+from .matchmaker import Matchmaker
+
+
+class Container:
+    def __init__(self, rating_difference, content):
+        self.rating = rating_difference
+        self.content = content
+
+
+@with_logger
+class TeamMatchMaker(Matchmaker):
+    """
+    Matchmaker for teams of varied size. Untested for higher than 4v4 but it should work
+
+    Overview of the algorithm:
+    1. list all the parties in queue by their average rating.
+    2. Take a party and select the neighboring parties alternating between lower
+    and higher until you have 8 players. If the next party to select would leave
+    you with more than 8 players you skip that party and try the next one.
+    3. you now have a list of parties that will be in your potential game. Distribute
+     them into two teams.
+     Start with the easiest cases: one party makes a full team already or one party
+     has n-1 players so you only need to find the best fitting single player.
+     If that is not the case perform the karmarkar-karp algorithm to get a good approximation for a partition
+    4. add this game to a games list
+    5. repeat 2. to 4. for every party.
+    6. you now have a list of potential games with minimal rating variation and minimal rating imbalance.
+    7. remove all games with match quality below threshold then sort by quality descending
+    8. pick the first game from the game list and remove all other games that contain the same players
+    9. repeat until the list is empty
```

Comment:
I think it's `O(n^2)`. Worst case we compare every game with the whole games list, so there we get n * n. (Maybe it is less than `O(n^2)` because the list will be shorter every time but I am not sure how this translates to O notation.) The player checking is `O(m)` where m is the player number in a game so should translate to `O(1)` for big n, right?